### PR TITLE
LibWeb: Get reference height from closest non-anonymous ancestor

### LINKED ISSUE #19052: LibWeb: Percentage height doesn't work correctly with anonymous block wrappers
For the following example:

```
<!DOCTYPE html>

<head>
    <style>
        .foo {
            width: 1280px;
            height: 800px;
        }

        .foo div {
            height: 600px;
        }

        .foo div img {
            max-height: 100%;
            float: right;
        }
    </style>
</head>

<body>

    <div class="foo">
        <div>
            <img
                src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAABLAAAAJYCAMAAABFOO8oAAABg1BMVEUAAAAXK0wXK00YJTwYLlEYLlIYLlQYMloYMlwYNmMZAAAZIjYZIzgZJDkZJj8ZLVEZMVkZQHcaNmQaOmsbMVQcJzwcNVsdPW4eK0QfNVYfNlwfQ3kfRXofR38gQHAhOF4hOWAiIiIiJCwiJS4iJi8iJzEiRnwjJy4jKz4jLEIjXK8jXLAkKjUkLDEkP2kkR3wkSIAkXbIkX7UkX7YlSYAlYLclYbkmMEYmSn0mYrwmZL8nZcEnZsInZsMoS30pZ8MqaMQracQva8UvbMUybsYzbsY0b8c1cMc2ccc5hwU6c8g6dMg6hwU7dMk7iQU7igU8dck8jQU9dslAlgZKlxVShc9Thc9YidFYzwha0ghbZH1b1Qhd2QleZ4Je3Qlf3glg4glh5Ali5gljkdRkkdRlbIdm5xBpcIhy6SJ00zR1ndl4n9p42jZ95Th/pNyB6zqDp92Q3luRseCY6mCb72Khq9KrsNGrsdOsstOts9OvtNOv8oK3AAC4vNi9wdu/w9zJAAAM9N5KAAAKzElEQVR42u3dzWscdRjA8Y1pG7EGtW9Ei6XRpj2poEfRWz0J6qGgBxHyJ3jw1EMOPfWP6FUErV60J0HISbEFWxQ1h0ix9gVtxaQx6TaJwjy/wvyc2c2msbuZ/XwuD7O72axZ+XYODzMjky2A7eEhfwJAsAAECxAsAMECECxAsAAEC0CwAMECECwAwQIEC0CwAAQLECwAwQIQLECwAAQLQLAAwQIQLADBAgQLQLAABAsQLADBAgQLQLAABAsQLADBAhAsQLAABAtAsADBAhAsAMECBAtAsAAECxAsAMECECxAsAAEC0CwAMECECwAwQIEC0CwAAQLECwAwQIEC0CwAAQLECwAwQIQLECwAAQLQLAAwQIQLADBAgQLQLAABAsQLADBAhAsQLAABAtAsADBAhAsAMECBAtAsAAECxAsAMECBAtAsAAECxAsAMECECxAsAAEC0CwAMECECwAwQIEC0CwAAQLECwAwQIQLECwAAQLQLAAwQIQLADBAgQLQLAABAsQLADBAgQLQLAABAsQLADBAhAsQLAABAtAsADBAhAsAMECBAtAsAAECxAsAMECECxAsAAEC0CwAMECECwAwQIEC0CwAAQLECwAwQIEC0CwAAQLECwAwQIQLECwAAQLQLAAwQIQLADBAgQLQLAABAsQLADBAhAsQLAABAtAsADBAhAsAMECBAtAsAAECxAsAMECBAtAsAAECxAsAMECECxAsAAEC2CTdvgT0NM/bX/FPB5ztBgX43BnzLsxl2Iulx9ux5yIuegPjDMsQLAABAtAsADBAhAsAMECtjF7WFQ7EPNyzNdj7q9++e3e3v1ozK9iXi2/+7K/P86wAMECECwAwQIEC0CwAP41MulvQMnfxXgjDh+P+XDNy68U42Yc7om5HvNgzFhU+DYO12KuxpyK+Wn54b2+DpxhAYIFIFgAggUIFoBgAYIFsB24vMzQ2xVzJea75f8zdmUvv5Id/16M2N66d5uv9WzOFyMtVl2r+TBvFWMuDr+JeTjmHV+XMywAwQIQLECwAAQLQLAAwQIYRK6HNbTSgtVYzCPVL0sXuEoXvEpXskr39WoXY3VrPtRU+TDtY30fM91jzD6WMywAwQIQLECwAAQLQLAAwQIYKPawhlZaqHppYy9bjNnOHn8g0nrW2ZhP+PqcYQEIFoBgAYIFIFgAggUIFsAgsYc1dNIdBN/MHr9RPlyOuRRztZ+fObtMVuvjmHt9nc6wAAQLQLAAwQIQLADBAhpthz/B0DhQjKc7P92aL8Zjcbi75uXXevvl2WLCePb0Qsy5jb3b8Zg/FGPZl+sMC0CwAAQLECwAwQIQLECwAPrKHtbwuFyMg3GYLieT9q9+iblUGq2JmGl9q2bnqW4ta6r64bR3NZ7NV4uxEodfZz+Wr2/9Vow9vlxnWACCBSBYgGABCBaAYAGCBdBXbvPVfOmKVrGttK/mZbdi3t3SXz7V28tj0Sr2q1pPxczXtpJfi3Hxf/noOMMCECxAsAAEC0CwAMECECyAjXI9rOa7HvNI9dPpwlcPdP9qruZlF8qH4zU/nvay4uaJ6WJc+3zZzrAABAtAsADBAhAsAMECBAugL+xhNd9IzNXyv1GHYrZj3ij/VL4CNbcln6Xbu3RZ33o05kL54VFfsjMsAMECECxAsAAEC0CwAMEC6Ct7WM03Uj6M2xO21rL/BdI9AOOSVOOb+2VTW/qypO7DrBfDHpYzLADBAhAsQLAABAtAsICGstbQfDvLhzezf6vS7b3u7/oxUw/kP+VqdnynGC/H4SVftjMsAMECECxAsAAEC0CwAMEC6At7WM2XLi8T15NZjsNb2csmYsZlZhZ6+yX5FtdUl+fvT9odizuXtX3JzrAABAtAsADBAhAsAMECBAugr+xhDY+4J9ZKHOaLVrPlwxO9vXnau5rb2Mvuk8UrZ1gAggUgWIBgAQgWgGABggUwUOxhNV9cBytdROrL7OlXsuPpYqSFqtku7z7d+emavazxmAvVj+cPp1sr7o65VH561JfsDAtAsAAECxAsAMECECxAsAD6yh5W86UbEa6UH077V/mi1Zne3j1enq9jdbtPYSbby6pb01qv/um0Wjbhy3aGBSBYAIIFCBaAYAEIFtAs1hqa74+YcZmZR+LwfKt8/EH1T8/09ss2eR+vhc5Pp3WG1eqn7/qSnWEBCBaAYAGCBSBYAIIFCBZAX9nDGh7ny4f53tVM9eFM9dO58ZrHd5YP2zEXOv9Yvpa11vmXj/lynWEBCBaAYAGCBSBYAIIFCBZAX9nDGjo11726d12spfLDp2OejHkq+7G4W9iFOHwxe7rd+cPkC1dxd7B0Va0fY75Q8+OLxUi391r17TrDAhAsAMECBAtAsAAECxAsgL4YmfQ3aLrrMWMB63T2dM1eVnrZ+zXvmvaxpotxJg5jLavX+xPOVT98O+ZyzOez5z8qxmFfsjMsAMECECxAsAAEC0CwAMEC6CvXwyIzU4y661+dzI7PVL9LWqzqcR9rvRgrcVi3fxXXwWodK78cZ1gAggUgWIBgAQgWgGABDWOtYejU3eZrJubJ3t4uLi9Tc32Yja43PBvzu2Ic6/JLz8U85Ot0hgUgWACCBQgWgGABCBYgWACDwB5W8y11fnomZs3+Vf5wfrmZWLSajcMTMffHvFT9rj/FbMdM9/N6rvNn/SSm/StnWACCBSBYgGABCBaAYAGCBTBQ7GENj5nSqN2/yhatul0ea7Z8eC17eiJmLF5NxuGfMeO2Xv+5j1fms5jPxLzj23SGBSBYAIIFCBaAYAEIFiBYAINkZNLfYFjMlw/rFqxq9rDy62BNV//0zzFXY6aNqeVirHf5jNk+1tmYk9m74QwLQLAABAsQLADBAhAsQLAABoo9rKFzPWa6XWGXC16dqnn8vZixcLVePrzn4sY+VNq/WizGuTh0/0GcYQGCBSBYAIIFCBaAYAGCBbAt2MMaWvO9vfydmB9WP/325j5E7F21voh5tBgrvh6cYQGCBSBYAIIFCBaAYAG0rDWw4f9TYqar07wWc60Y6bIy+eVlRsvz8zgci/lkzY+BMyxAsAAEC0CwAMECECwAwQK2E3tYgDMsAMECBAtAsAAECxAsAMECECxAsAAEC0CwAMECECwAwQIEC0CwAAQLECwAwQIQLECwAAQLQLAAwQIQLADBAgQLQLAAwfInAAQLQLAAwQIQLADBAgQLQLAABAsQLADBAhAsQLAABAtAsADBAhAsAMECBAtAsAAECxAsAMECECxAsAAEC0CwAMECECxAsAAEC0CwAMECECwAwQIEC0CwAAQLECwAwQIQLECwAAQLQLAAwQIQLADBAgQLQLAABAsQLADBAhAsQLAABAtAsADBAhAsQLAABAtAsADBAhAsAMECBAtAsAAECxAsAMECECxAsAAEC0CwAMECECwAwQIEC0CwAAQLECwAwQIQLECwAAQLQLAAwQIQLECwAAQLQLAAwQIQLADBAgQLQLAABAsQLADBAhAsQLAABAtAsADBAhAsAMECBAtAsAAECxAsAMECECxAsAAEC0CwAMECECxAsAAEC0CwAMECECwAwQIEC0CwAAQLECwAwQIQLECwAAQLQLAAwQIQLADBAgQLQLAABAsQLADBAhAsQLAABAtAsADBAhAsQLAABAtAsADBAhAsAMECBAtAsAAEC2i8fwDFkO6lWFprKQAAAABJRU5ErkJggg==">
            <p>Test</p>
        </div>
    </div>

</body>

</html>
```

Ladybird only renders the paragraph:
ladybird

while Chrome / Firefox render the image too:
chrome

The problem appears to be that we resolve the percentage height specified for the image with respect to the immediate containing block, which doesn't have a defined height and we ultimately end up with a height of 0:

```
        ...
        BlockContainer <div> at (8,16) content-size 1280x600 [0+0+0 1280 0+0+0] [0+0+0 600 0+0+0] children: not-inline
          BlockContainer <(anonymous)> at (8,16) content-size 1280x0 [0+0+0 1280 0+0+0] [0+0+0 0 0+0+0] children: inline
            TextNode <#text>
            ImageBox <img> at (88,16) content-size 1200x0 floating [0+0+0 1200 0+0+0] [0+0+0 0 0+0+0] children: not-inline
        ...
```

This is part of the reason why the [fundamental CSS comprehension example on MDN](https://mdn.github.io/learning-area/css/introduction-to-css/fundamental-css-comprehension-finished/index.html) renders incorrectly on Ladybird.

### PR DESCRIPTION
**Pull request by axgallo**
Ignore anonymous block boxes when resolving percentage weights that would refer to them, per the CSS 2 visual formatting model specification. This fixes the case when we create an anonymous block between an image which uses a percentage height relative to a parent which specifies a definite height.

Fixes #19052.

### CODE REVIEW

**Comment by awesomekling** [comment]
Can you refer to the specification text that tells us to behave this way?

**Comment by axgallo** [comment]
Not sure, is there a part of the spec which describes when we should create wrapper blocks? In the original DOM, the height is defined in the containing block, but once we wrap the image in an anonymous block we attempt to take the height from that (and it doesn't have one defined).

**Comment by awesomekling** [comment]
> Not sure, is there a part of the spec which describes when we should create wrapper blocks? In the original DOM, the height is defined in the containing block, but once we wrap the image in an anonymous block we attempt to take the height from that (and it doesn't have one defined).

"anonymous box" is defined in CSS-DISPLAY-3: https://www.w3.org/TR/css-display-3/#anonymous

For height sizing, https://www.w3.org/TR/css-sizing-3/ has a lot of stuff, and https://www.w3.org/TR/CSS22/ fills in the blanks sometimes.

The current fix here looks like an arbitrary hack to make the testcase work. We need to find the real "spec reason" to have this behavior. Alternatively, if this *isn't* in the spec, we should look at what other engines do and behave the same way :)

**Comment by axgallo** [comment]
Thanks, I'll review that section of the spec and I'll update the PR accordingly. It looks that it is specified to a sufficient extent to allow a principled fix.

**Comment by axgallo** [comment]
Last paragraph of https://www.w3.org/TR/CSS22/visuren.html#anonymous-block-level says the following:

"Anonymous block boxes are ignored when resolving percentage values that would refer to it: the closest non-anonymous ancestor box is used instead. For example, if the child of the anonymous block box inside the DIV above needs to know the height of its containing block to resolve a percentage height, then it will use the height of the containing block formed by the DIV, not of the anonymous block box."

I've adjusted the patch to do that.

**Comment by awesomekling** [comment]
Aha! There we go, that's a great discovery. We probably need to account for this in more places, but now we have the right knowledge and helper function for it :)

File: Userland/Libraries/LibWeb/Layout/FormattingContext.cpp [hunk 1]
```diff
@@ -1358,7 +1358,8 @@ CSS::Length FormattingContext::calculate_inner_width(Layout::Box const& box, Ava
 
 CSS::Length FormattingContext::calculate_inner_height(Layout::Box const& box, AvailableSize const&, CSS::Size const& height) const
 {
-    auto height_of_containing_block = m_state.get(*box.containing_block()).content_height();
+    auto height_of_containing_block = m_state.get(*box.non_anyonymous_containing_block()).content_height();
+    ;
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/Layout/FormattingContext.cpp:1362]: Stray `;` here.

Comment (axgallo) [reply Userland/Libraries/LibWeb/Layout/FormattingContext.cpp]: Thanks, removed it.

**Pull request merged**